### PR TITLE
Set the `in-rust-tree` feature for all rust-analyzer{-proc-macro-srv} steps

### PR DESCRIPTION
**Pull request by Veykril**
Some context: This came up in https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/libraryfying.20match.20checking.3F where `test` and `check` behaved differently for rust-analyzer
r? @onur-ozkan

### CODE REVIEW

File: src/bootstrap/src/core/build_steps/check.rs [hunk 1]
```diff
@@ -386,7 +386,7 @@ impl Step for RustAnalyzer {
             cargo_subcommand(builder.kind),
             "src/tools/rust-analyzer",
             SourceType::InTree,
-            &["rust-analyzer/in-rust-tree".to_owned()],
+            &["in-rust-tree".to_owned()],
```
Comment (Veykril) [inline src/bootstrap/src/core/build_steps/check.rs:389]: I changed these prefixed ones given we didn't prefix the `sysroot-abi` feature either in the other places, and I was kind of confused that these worked in the first place

**Review by onur-ozkan — COMMENTED** [review]
I guess you missed this part https://github.com/rust-lang/rust/blob/8c2c233e6b94ca532139d7af5d7dc4979f66f319/src/bootstrap/src/core/build_steps/tool.rs#L650

**Comment by nnethercote** [comment]
I tested and it fixes my original problem, thanks!

**Comment by Veykril** [comment]
I don't understand the issue here 😕 The crates are listed in the `Cargo.toml`

**Comment by bjorn3** [comment]
`rust-analyzer-proc-macro-srv` is marked as `ToolStd` and thus assumed to not depend on the compiler libraries. Try changing it to `ToolRustc`.

**Comment by onur-ozkan** [comment]
Lol didn't expect that. @bors r-

**Comment by nnethercote** [comment]
Any idea what the problem is? It would be good to get this problem fixed.

**Comment by onur-ozkan** [comment]
> Any idea what the problem is?

When `RustAnalyzerProcMacroSrv` is switched to `ToolRustc`, it compiles the compiler as well. Currently `RustAnalyzerProcMacroSrv` is compiled even with the default configuration, changing the default bootstrap behavior and causing bootstrap tests to fail. #119124 will fix the issue.

@rustbot blocked

**Comment by onur-ozkan** [comment]
@Veykril Can you rebase to get #119124 fix on this PR?

**Comment by onur-ozkan** [comment]
Seems that we need to perform specific work on the `i686-mingw` target to build `rust-analyzer-proc-macro-srv` tool. I currently don't have a development environment for `i686-mingw`. Due to other tasks that I have already assigned, I can only set up the environment and start working on this in ~3 weeks or so. If you have more time available, feel free to fix this before me.

**Comment by nnethercote** [comment]
It is currently impossible to change the public API of the `rustc_lexer` crate because of this issue. It has now gone through several rounds of attempts to fix it, none of which has fully succeeded. This is a problem. In particular, I need to do #119172 or something similar soon, and this issue is blocking that.

Why is the issue so complicated? I am not happy that this rustc/rust-analyzer code sharing is preventing rustc work from occurring.

My understanding is that the trend is towards more sharing of code between rustc and rust-analyzer. This issue is not giving me confidence in that trend.

cc @rust-lang/compiler

**Comment by onur-ozkan** [comment]
> It has now gone through several rounds of attempts to fix it, none of which has fully succeeded.

> Why is the issue so complicated?

They were all different problems. 

> I am not happy that this rustc/rust-analyzer code sharing is preventing rustc work from occurring.

Same here. I think it should be the other way around — testing rust-analyzer with rustc in the rust-analyzer CI pipelines and making it compatible either on the upstream or in rust-analyzer when needed (this way rustc will not be blocked). I think I suggested this couple times in zulip channels.

**Comment by lnicola** [comment]
Testing it with the rustc crates won't catch most of these problems, as I mentioned the last time. We'd have to clone the whole compiler repository, merge the changes, and run bootstrap to find them.

**Comment by onur-ozkan** [comment]
>  We'd have to clone the whole compiler repository, merge the changes, and run bootstrap to find them.

Cloning compiler repository and replacing RA shouldn't be hard I think?

**Comment by lnicola** [comment]
It's not easy either, because 1. we sometimes get merge conflicts, 2. we use standard runners, not the fancy xl ones, 3. we don't test all configurations (certainly not i686-mingw), 4. we don't know much about the rustc CI setup, and 5. we don't know much about bootstrap.

**Comment by HKalbasi** [comment]
The problem here is a one time problem, which is happening now because the `in-rust-tree` thing is relatively new and #119172 is effectively the first PR that is actually using it. So this kind of problem hopefully won't happen regularly. For unblocking #119172 we may want to temporarily disable r-a in CI or something like that to fix this one time problem without time pressure (but ideally we merge this PR).

But if the r-a related changes in #119172 are hard or are blocking the development (which I don't think so) that is a problem which make us reconsider running r-a in CI.

**Comment by onur-ozkan** [comment]
> For unblocking https://github.com/rust-lang/rust/pull/119172 we may want to temporarily disable r-a in CI or something like that to fix this one time problem without time pressure (but ideally we merge this PR).

I haven't tested, but this patch should temporarily disable RA on the CI for the failing target.

```diff
diff --git a/src/bootstrap/src/core/build_steps/tool.rs b/src/bootstrap/src/core/build_steps/tool.rs
index 8e3941dbeda..1ac635e153b 100644
--- a/src/bootstrap/src/core/build_steps/tool.rs
+++ b/src/bootstrap/src/core/build_steps/tool.rs
@@ -634,6 +634,11 @@ fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
     }
 
     fn make_run(run: RunConfig<'_>) {
+        // FIXME: https://github.com/rust-lang/rust/pull/118861#issuecomment-1868251547
+        if build_helper::ci::CiEnv::is_ci() && run.build_triple() == "i686-pc-windows-gnu" {
+            return;
+        }
+
         run.builder.ensure(RustAnalyzer {
             compiler: run.builder.compiler(run.builder.top_stage, run.builder.config.build),
             target: run.target,
@@ -682,6 +687,11 @@ fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
     }
 
     fn make_run(run: RunConfig<'_>) {
+        // FIXME: https://github.com/rust-lang/rust/pull/118861#issuecomment-1868251547
+        if build_helper::ci::CiEnv::is_ci() && run.build_triple() == "i686-pc-windows-gnu" {
+            return;
+        }
+
         run.builder.ensure(RustAnalyzerProcMacroSrv {
             compiler: run.builder.compiler(run.builder.top_stage, run.builder.config.build),
             target: run.target,
```

**Comment by Veykril** [comment]
Pushed the temporary disabling fix.

> Same here. I think it should be the other way around — testing rust-analyzer with rustc in the rust-analyzer CI pipelines and making it compatible either on the upstream or in rust-analyzer when needed (this way rustc will not be blocked). I think I suggested this couple times in zulip channels.

The entire reason why we want the rust-lang/rust CI testing rust-analyzer with the in-tree-deps is so that the r-a team does not have to chase down commits that made things fully incompatible again (like when the rustc-abi crate completely removed the rust-analyzer needed parts as well as making it no longer compile on stable). Having r-a run this in its own CI makes that pointless, as at best we just have a CI that randomly starts failing when things change in the rustc repo. 

Obviously this comes at the cost of the rustc contributors having to touch the rust-analyzer parts when they change exposed API but in that case I am happy to help out were needed.

**Comment by onur-ozkan** [comment]
The problem isn't platform specific seems like. We need to update the checks with something like `target.contains("windows-gnu")`. (feel free to r=me after that)

**Comment by nnethercote** [comment]
> Obviously this comes at the cost of the rustc contributors having to touch the rust-analyzer parts when they change exposed API

This is how things already work for rustfmt and clippy, and it works fine. Hopefully r-a can get to a similarly reliable state soon. Thanks for all the responses here!

**Comment by onur-ozkan** [comment]
> This is a problem. In particular, I need to do https://github.com/rust-lang/rust/pull/119172 or something similar soon, and this issue is blocking that.

How does this PR helps you for that? I am doing some debugging on CI (because my current dev-environment isn't enough for testing this) at #119279, which I cherry-picked your commit from #119172 and the first commit from this PR. The result is `mingw-check` is still failing:


```sh
  error[E0004]: non-exhaustive patterns: `EscapeError::NulInCStr` not covered
     --> crates/parser/src/lexed_str.rs:343:11
      |
  343 |     match error {
      |           ^^^^^ pattern `EscapeError::NulInCStr` not covered
      |
  note: `EscapeError` defined here
     --> /checkout/compiler/rustc_lexer/src/unescape.rs:16:1
      |
  16  | pub enum EscapeError {
      | ^^^^^^^^^^^^^^^^^^^^
  ...
  63  |     NulInCStr,
      |     --------- not covered
      = note: the matched value is of type `EscapeError`
  help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
      |
  375 ~         EscapeError::MultipleSkippedLinesWarning => "",
  376 ~         EscapeError::NulInCStr => todo!(),
      |
  
  For more information about this error, try `rustc --explain E0004`.
  error: could not compile `parser` (lib) due to 1 previous error
  warning: build failed, waiting for other jobs to finish...
  Build completed unsuccessfully in 0:03:30
    local time: Sun Dec 24 15:29:34 UTC 2023
    network time: Sun, 24 Dec 2023 15:29:34 GMT
  Error: Process completed with exit code 1.
```

**Comment by onur-ozkan** [comment]
mingw linking fails due to passing both `sysroot-abi` and `in-rust-tree` features at the same time for `src/tools/rust-analyzer/crates/proc-macro-srv-cli` which doesn't happen when passing only one of them. It might be worth to re-check how these features are managed within `rust-analyzer`. I would continue digging on the link-time problem once I get more/enough time.

**Comment by nnethercote** [comment]
> How does this PR helps you for that? I am doing some debugging on CI (because my current dev-environment isn't enough for testing this) at https://github.com/rust-lang/rust/pull/119279, which I cherry-picked your commit from https://github.com/rust-lang/rust/pull/119172 and the first commit from this PR.

With #119279 locally I could run `x.py test` successfully but `x.py check` was failing. If I fixed the `x.py check` failure, then `x.py test` started failing. Because `x.py check` was using the locally modified rustc_lexer but `x.py test` was not. At least, that's my understanding. This is the problem that [started everything](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/rust-lang.2Frust.20repo.20syncing).

**Comment by onur-ozkan** [comment]
I dig into the problem more today and found out that these blocks

https://github.com/rust-lang/rust/blob/f8fe51714470dcd10b4787df27fb59b3a2775026/src/tools/rust-analyzer/crates/proc-macro-srv-cli/src/main.rs#L39-L42

are the ones which causes linking error on windows-gnu when "in-rust-tree" feature is enabled (confirmed both locally and on CI https://github.com/rust-lang/rust/actions/runs/7323280040/job/19945709276?pr=119279#step:26:7919 with https://github.com/rust-lang/rust/pull/119279/commits/8c88c1d7417231afe1276bd7eafa617c6f380abc). Now we know the entrypoint if the problem, but since I am not so experienced on the RA codebase, it would probably be much better if someone from RA team can work on this problem.

**Comment by Veykril** [comment]
I fail to see the linking error in that CI log (in fact I fail to see the actual error at all). I won't look into this until new years since I'm taking a break now.

**Comment by onur-ozkan** [comment]
> I fail to see the linking error in that CI log

There is no error. I meant https://github.com/rust-lang/rust/commit/8c88c1d7417231afe1276bd7eafa617c6f380abc allows us to build `rust-analyzer-proc-macro-srv` on *-windows-gnu without any issue. So the commented lines in the commit somehow failing the `rust-analyzer-proc-macro-srv` builds when `sysroot-abi` and `in-rust-tree` features are enabled on *-windows-gnu.

**Comment by Veykril** [comment]
I have no idea what could cause those linking issues. Those have nothing to do with rust-analyzer itself, but rather some misconfiguration from bootstrap (that is the build step for the server missing something, or some bug in bootstrap). I changed some things around now within r-a itself but I doubt that will have an effect.

@bors r=onur-ozkan

**Comment by Veykril** [comment]
Well I don't know what could cause this, given the symbols it fails to link are normal rust symbols, and there is nothing special going with the feature flags in `proc-macro-srv` itself, aside from `sysroot-abi` configuring the entire crate away if disabled. Even then, this should give a name res diagnostic if anything, not a linker error. So I highly doubt this is reproducible with plain cargo, though I am open to test if anyone can tell me how to.

**Comment by onur-ozkan** [comment]
> I am quite sure the problem will happen outside bootstrap (using cargo directly) too.

`cargo -v -V`:

```sh
onuro@Onur MINGW64 ~/rust/src/tools/rust-analyzer/crates/proc-macro-srv-cli
$ cargo -v -V
cargo 1.77.0-nightly (ac6bbb332 2023-12-26)
release: 1.77.0-nightly
commit-hash: ac6bbb33293d8d424c17ecdb42af3aac25fb7295
commit-date: 2023-12-26
host: x86_64-pc-windows-gnu
libgit2: 1.7.1 (sys:0.18.1 vendored)
libcurl: 8.5.0-DEV (sys:0.4.70+curl-8.5.0 vendored ssl:Schannel)
os: Windows 10.0.22000 (Windows 11 Professional) [64-bit]
```

`cargo build --features "sysroot-abi in-rust-tree"`: [complete build log](https://gist.github.com/onur-ozkan/e5f92b9eca16d63c4a6d7514ccec8b18)

> I am open to test if anyone can tell me how to.

Simply follow the mingw instructions at https://github.com/rust-lang/rust?tab=readme-ov-file#mingw on Windows, install a nightly toolchain with rust-src component and build `proc-macro-srv-cli` with "sysroot-abi in-rust-tree" features.

**Comment by Veykril** [comment]
I am not getting very far with that
```
lukas@DESKTOP-BR3A5QS MINGW64 /c/workspace/rust/rust/src/tools/rust-analyzer/crates/proc-macro-srv-cli
$ cargo build --features "sysroot-abi in-rust-tree"
   Compiling proc-macro-srv-cli v0.0.0 (C:\workspace\rust\rust\src\tools\rust-analyzer\crates\proc-macro-srv-cli)
error: Error calling dlltool 'dlltool.exe': program not found

error: could not compile `proc-macro-srv-cli` (bin "rust-analyzer-proc-macro-srv") due to 1 previous error

```

```
lukas@DESKTOP-BR3A5QS MINGW64 /c/workspace/rust/rust/src/tools/rust-analyzer/crates/proc-macro-srv-cli
$ cargo -v -V
cargo 1.77.0-nightly (ac6bbb332 2023-12-26)
release: 1.77.0-nightly
commit-hash: ac6bbb33293d8d424c17ecdb42af3aac25fb7295
commit-date: 2023-12-26
host: x86_64-pc-windows-gnu
libgit2: 1.7.1 (sys:0.18.1 vendored)
libcurl: 8.5.0-DEV (sys:0.4.70+curl-8.5.0 vendored ssl:Schannel)
os: Windows 10.0.19045 (Windows 10 Home) [64-bit]
```

Had a lot more problems setting that up than "simply" as well unfortunately (which I expected, mingw is a pain). Will brainstorm what could be going wrong here tomorrow but there is really nothing special in r-a going on...

**Comment by mati865** [comment]
That doesn't seem right, what are the outputs from:
`which gcc`, `which dlltool`, `which ld`?

**Comment by Veykril** [comment]
```
lukas@DESKTOP-BR3A5QS MINGW64 ~
$ which gcc
/mingw64/bin/gcc

lukas@DESKTOP-BR3A5QS MINGW64 ~
$ which dlltool
/mingw64/bin/dlltool

lukas@DESKTOP-BR3A5QS MINGW64 ~
$ which ld
/mingw64/bin/ld

```

**Comment by Veykril** [comment]
Okay good news is I can build it as well now (was missing the llvm-tools component), bad news it compiles successfully on my machine

**Comment by Veykril** [comment]
(rebased)

**Comment by onur-ozkan** [comment]
Last fix attemps seems to fixed the issue. I can no longer reproduce the problem with x and cargo.

**Comment by Veykril** [comment]
Lovely, I think it was a misplaced `extern crate rustc_driver` then (I recall moving that). Then this is good to go again.

**Comment by onur-ozkan** [comment]
> I think it was a misplaced extern crate rustc_driver then (I recall moving that)

Could be. Or, maybe rebasing fixed the problem with the RA changes (I see there are several changes on the part which was causing this error).

Will r+ the PR once CI is green.

File: src/doc/reference
Comment (onur-ozkan) [inline src/doc/reference:1]: I guess these submodule changes pushed accidentally ?

Comment (Veykril) [reply src/doc/reference:1]: Yes, rebasing always seems to mess with the submodules ...

Comment (Veykril) [reply src/doc/reference:1]: Fixed

**Comment by Veykril** [comment]
Made when resolving a conflict while rebasing earlier

**Pull request merged**